36535 implement contenteditable plaintext only (#36751)
mdn/content: files/en-us/web/html/global_attributes/contenteditable/index.md

Elements that are made editable, and therefore interactive, by using the `contenteditable` attribute can be focused. They participate in sequential keyboard navigation. However, elements with the `contenteditable` attribute nested within other `contenteditable` elements are not added to the tabbing sequence by default. You can add the nested `contenteditable` elements to the keyboard navigation sequence by specifying the `tabindex` value ([`tabindex="0"`](/en-US/docs/Web/HTML/Global_attributes/tabindex)).

> If content is pasted into an element with `contenteditable="true"` all of the formatting is retained. If content is pasted into an element with `contenteditable="plaintext-only"` all of the formatting is removed.

Reviewer: ```suggestion
If content is pasted into an element with `contenteditable="true"`, all formatting is retained. If content is pasted into an element with `contenteditable="plaintext-only"`, all formatting is removed.
```

## Examples

### Pasting content into contenteditable

> This example has 2 {{HTMLElement("div")}}s both with `contenteditable`, the first with the value `true` and the second with the value `plaintext-only`. Copy the content below and paste it into each of the `div`s to see the different effects.

Reviewer: ```suggestion
This example has two {{HTMLElement("div")}} elements with `contenteditable`, the first with the value `true` and the second with the value `plaintext-only`. Copy the content below and paste it into each `div` to see their effects.
```

#### HTML

```html hidden
<h2>Content to copy</h2>
<p class="instructions">

>   Copy all of the text in the block below and try pasting it into each of the
>   blocks below.

Reviewer: ```suggestion
  Copy all the text in the block below and paste it into each of the contenteditable
  blocks to compare the results.
```

</p>
<section class="copying">
  <div class="copy">
    <p>

>       This is a paragraph with <strong>Bold</strong> <em>Italic</em> and
>       <span class="red">red text</span>. Below is an ordered list:

Reviewer: ```suggestion
      This is a paragraph containing <strong>Bold</strong>, <em>Italic</em>, and
      <span class="red">red</span> text, followed by an ordered list:
```

    </p>
    <ol>
      <li>Step one</li>
      <li>Step two</li>

>       <li>Step Three</li>

Reviewer: ```suggestion
      <li>Step three</li>
```

    </ol>
  </div>
</section>
```

```html

> <h2>Pasting areas</h2>

Reviewer: Can we reduce the space between "Pasting areas" and "contenteditable="true""? Maybe change the h3 to a para tag?

Author: done I adjusted the margins

<section class="pasting">
  <div class="wrapper">
    <h3>contenteditable="true"</h3>
    <div contenteditable="true"></div>
  </div>
  <div class="wrapper">
    <h3>contenteditable="plaintext-only"</h3>
    <div contenteditable="plaintext-only"></div>
  </div>

> </section>

Reviewer: I had only checked in Chrome earlier.

I tested in Firefox now:
1. I am sometimes getting this inner box boundary on pasting (even in Chrome):

![Screenshot 2024-11-12 at 12 53 36 PM](https://github.com/user-attachments/assets/c7f54e42-60af-4465-8be5-1c5f00ea9ff1)

2. I'm unable to paste the text in the second box with `contenteditable="plaintext-only"`.
I've set `dom.element.contenteditable.plaintext-only.enabled` to `true`.
Is this working alright at your end?

3. I missed it on the Experimental features page - the last string in the preference name should be `enabled`. It's currently `dom.element.contenteditable.plaintext-only.enable`. Can you also add that fix to this PR?

Author: 1. that is becase you are copying all the formatting and this is as expected
2. This is because you are on firefox 132 if you use Beta 133 or developer edition, this works fine
3. fixing this now

Author: Done

Reviewer: I was using the wrong version of Firefox :-/ It is working in Nightly.
Thank you!

```

```css hidden
.copying {
  font-family: Georgia, serif;
  margin: 1rem;
  padding: 1rem;
  border: solid black 1px;
}
.red {
  color: red;
}
.pasting {
  display: flex;
  flex-direction: row;
  gap: 1rem;
  width: 100%;
  .wrapper {
    flex: 1 1;
    /* margin: 1rem; */
    padding: 1rem;
    /* border: solid tomato 1px; */
    border-radius: 1rem;
    /* background-color: lightgrey; */
  }
  h3 {
    font-family: monospace;
  }
  [contenteditable] {
    min-height: 3rem;
    border: solid 1px;
    padding: 0.5rem;
    background-color: whitesmoke;
  }
  [contenteditable="true"] {
    caret-color: blue;
  }
  [contenteditable="plaintext-only"] {
    caret-color: red;
  }
}
```

> {{EmbedLiveSample("Pasting_Content_into_contenteditable", 400, 500)}}

Reviewer: The height can be increased to preempt the iframe height increase after content is pasted

Author: done
